File: 0501-1000/0653. Two Sum IV - Input is a BST.java
// Easy
// DFS, Hash Table, Two Pointers
// O(n)
// https://leetcode.cn/problems/two-sum-iv-input-is-a-bst/

import java.util.*;

class Solution {
    // two sum -> hash table, BST -> DFS
    private Set<Integer> vis = new HashSet<>();
    private int k;

    public boolean findTarget(TreeNode root, int k) {
        this.k = k;
        return dfs(root);
    }

    private boolean dfs(TreeNode root) {
        if (root == null) {
            return false;
        }

        // first to check if set contains the k - root.val
        if (vis.contains(k - root.val)) {
            return true;
        }
        vis.add(root.val);
        return dfs(root.left) || dfs(root.right);
    }
}

// followup:如果不能用額外空間？
// 可以用BST的特性：中序遍歷為有序數組，來找到目標值 -> Two pointers
class Solution2 {
    public boolean findTarget(TreeNode root, int k) {
        List<Integer> nums = new ArrayList<>();
        inOrder(root, nums);

        // sorted array
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int sum = nums.get(left) + nums.get(right);
            if (sum == k) {
                return true;
            } else if (sum < k) {
                left++;
            } else {
                right--;
            }
        }
        return false;
    }

    private void inOrder(TreeNode root, List<Integer> nums) {
        if (root == null) return;
        inOrder(root.left, nums);
        nums.add(root.val);
        inOrder(root.right, nums);
    }
}

File: 0501-1000/0653. Two Sum IV - Input is a BST.py
# Time:O(n), Space:O(n)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        vis = set()

        def dfs(node):
            if not node:
                return False
            if k - node.val in vis:
                return True
            vis.add(node.val)
            return dfs(node.left) or dfs(node.right)

        return dfs(root)
